File: frontend/src/types/chart.ts
/* ── Chart type definitions mirroring backend API ─────── */

export type ChartType =
  | 'natal'
  | 'transit'
  | 'progressed'
  | 'solar_return'
  | 'synastry_composite'
  | 'event';

export type BodyKey =
  | 'sun' | 'moon' | 'mercury' | 'venus' | 'mars'
  | 'jupiter' | 'saturn' | 'uranus' | 'neptune' | 'pluto'
  | 'north_node' | 'south_node' | 'chiron'
  | 'asc' | 'mc' | 'dsc' | 'ic'
  | 'lilith' | 'part_of_fortune';

export type AspectType =
  | 'conjunction' | 'sextile' | 'square' | 'trine' | 'opposition'
  | 'quincunx' | 'semi_sextile' | 'semi_square' | 'sesquiquadrate';

export type HouseSystem = 'P' | 'W' | 'K' | 'E' | 'R' | 'C' | 'V';

export type SignName =
  | 'aries' | 'taurus' | 'gemini' | 'cancer'
  | 'leo' | 'virgo' | 'libra' | 'scorpio'
  | 'sagittarius' | 'capricorn' | 'aquarius' | 'pisces';

export type Dignity =
  | 'domicile' | 'exaltation' | 'detriment' | 'fall' | 'peregrine';

export type RenderMode = 'replace' | 'overlay' | 'split';

/* ── Interfaces ──────────────────────────────────────── */

export interface ChartLocation {
  latitude: number;
  longitude: number;
  time_zone: string;
  location_name?: string;
}

export interface ChartBody {
  body_key: BodyKey;
  longitude: number;
  latitude?: number;
  sign: SignName;
  sign_degree: number;
  house: number | null;
  speed: number;
  is_retrograde: boolean;
  dignity: Dignity | null;
}

export interface ChartHouse {
  house_number: number;
  cusps_longitude: number;
  sign: SignName;
  sign_degree: number;
}

export interface ChartAspect {
  body_a_key: BodyKey;
  body_b_key: BodyKey;
  aspect_type: AspectType;
  orb: number;
  is_applying: boolean | null;
  is_major: boolean;
}

export interface ChartMetadata {
  ephemeris_version?: string;
  calculation_library?: string;
  house_system?: string;
}

export interface ChartPayload {
  id: string;
  chart_type: ChartType;
  user_id?: string;
  calculation_date: string;
  location?: ChartLocation;
  house_system: HouseSystem;
  latitude?: number;
  longitude?: number;
  time_zone?: string;
  location_name?: string;
  title?: string;
  bodies: ChartBody[];
  houses: ChartHouse[];
  aspects: ChartAspect[];
  metadata: ChartMetadata;
  created_at?: string;
}

export interface ChartSummary {
  id: string;
  chart_type: ChartType;
  calculation_date: string;
  sun: ChartBody | null;
  moon: ChartBody | null;
  rising: ChartBody | null;
  key_aspects: ChartAspect[];
}

/* ── Transit types ───────────────────────────────────── */

export interface TransitEvent {
  date: string;
  exact_date: string;
  transiting_body: BodyKey;
  natal_body: BodyKey;
  aspect_type: AspectType;
  orb: number;
  is_stationing: boolean;
  is_entering_sign: boolean;
}

export interface RetrogradePeriod {
  body: BodyKey;
  station_retrograde: string;
  station_direct: string;
  affected_natal_houses: number[];
}

export interface TransitSnapshot {
  natal_chart_id: string;
  date_from: string;
  date_to: string;
  events: TransitEvent[];
  retrograde_periods: RetrogradePeriod[];
}

/* ── Synastry types ──────────────────────────────────── */

export interface SynastryHouseOverlay {
  body_key: BodyKey;
  in_partner_house: number;
  partner_chart_id: string;
}

export interface SynastryScore {
  emotional: number;
  communication: number;
  passion: number;
  commitment: number;
  overall: number;
}

export interface SynastryPayload {
  id: string;
  chart_a: ChartPayload;
  chart_b: ChartPayload;
  relationship_type?: string;
  inter_aspects: ChartAspect[];
  house_overlays: SynastryHouseOverlay[];
  composite_chart?: ChartPayload;
  score_summary?: SynastryScore;
  render_mode?: 'bi_wheel' | 'composite' | 'aspect_table';
}

/* ── Life Phase types ────────────────────────────────── */

export interface LifePhase {
  id: number;
  phase_key: string;
  title: string;
  start_date: string;
  end_date: string | null;
  dominant_transits: Record<string, unknown> | null;
  description: string | null;
}

File: backend/app/schemas/chart.py
"""Pydantic v2 schemas for the ``charts`` resource."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ======================================================================
# Request schemas
# ======================================================================

class ChartLocation(BaseModel):
    """Location data for chart calculation."""

    latitude: Decimal = Field(..., ge=-90, le=90, decimal_places=6)
    longitude: Decimal = Field(..., ge=-180, le=180, decimal_places=6)
    time_zone: str
    location_name: str | None = None


class ChartOptions(BaseModel):
    """Optional parameters for chart calculation."""

    bodies: list[str] | None = None
    aspects: list[str] | None = None
    orb_major: float = 8.0
    orb_minor: float = 2.0
    include_minor_aspects: bool = True


class ChartCreate(BaseModel):
    """Request body for ``POST /v1/charts``."""

    chart_type: str = Field(
        ...,
        pattern="^(natal|transit|progressed|solar_return|synastry_composite|event)$",
    )
    reference_chart_id: UUID | None = None
    calculation_date: datetime
    location: ChartLocation
    house_system: str = Field(default="P", pattern="^[PWKERCV]$")
    options: ChartOptions | None = None


# ======================================================================
# Response schemas
# ======================================================================

class ChartBodyResponse(BaseModel):
    """A single celestial body in a chart."""

    model_config = ConfigDict(from_attributes=True)

    body_key: str
    longitude: Decimal
    sign: str
    sign_degree: Decimal
    house: int | None = None
    speed: Decimal
    is_retrograde: bool
    dignity: str | None = None


class ChartHouseResponse(BaseModel):
    """A single house cusp in a chart."""

    model_config = ConfigDict(from_attributes=True)

    house_number: int
    cusps_longitude: Decimal
    sign: str
    sign_degree: Decimal


class ChartAspectResponse(BaseModel):
    """A single aspect between two bodies in a chart."""

    model_config = ConfigDict(from_attributes=True)

    body_a_key: str
    body_b_key: str
    aspect_type: str
    orb: Decimal
    is_applying: bool | None = None
    is_major: bool


class ChartResponse(BaseModel):
    """Full chart response matching ``GET /v1/charts/:id``."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    chart_type: str
    user_id: UUID
    calculation_date: datetime
    house_system: str
    latitude: Decimal
    longitude: Decimal
    time_zone: str
    location_name: str | None = None
    title: str | None = None
    bodies: list[ChartBodyResponse] = []
    houses: list[ChartHouseResponse] = []
    aspects: list[ChartAspectResponse] = []
    metadata: dict = {}
    created_at: datetime


class ChartSummaryResponse(BaseModel):
    """Lightweight chart summary (signs, key aspects)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    chart_type: str
    calculation_date: datetime
    sun: ChartBodyResponse | None = None
    moon: ChartBodyResponse | None = None
    rising: ChartBodyResponse | None = None
    key_aspects: list[ChartAspectResponse] = []
